File: 2022/02/25/moveZeroes.ts
/**
 * https://leetcode.com/problems/move-zeroes/
 * @author Andrew Jarombek
 * @since 2/25/2022
 */

function moveZeroes(nums: number[]): void {
    let slow = 0;
    let fast = 0;

    while (fast < nums.length) {
        if (nums[fast] != 0) {
            nums[slow] = nums[fast];
            slow++;
        }

        fast++;
    }

    while (slow < nums.length) {
        nums[slow] = 0;
        slow++;
    }
}

File: 2022/07/12/java/MoveZeroes.java
/**
 * https://leetcode.com/problems/move-zeroes
 * @author Andrew Jarombek
 * @since 7/12/2022
 */

public class MoveZeroes {
    public static void moveZeroes(int[] nums) {
        int i = 0;
        int j = 0;

        while (j < nums.length) {
            int val = nums[j];
            j++;

            if (val != 0) {
                nums[i] = val;
                i++;
            }
        }

        while (i < nums.length) {
            nums[i] = 0;
            i++;
        }
    }

    public static void main(String[] args) {
        int[] nums = new int[] {0, 1, 0, 2, 3};
        moveZeroes(nums);

        assert nums[0] == 1;
        assert nums[1] == 2;
        assert nums[2] == 3;
        assert nums[3] == 0;
        assert nums[4] == 0;
    }
}
